content/en/blog/_posts/2023-12-13-scheduler-queueinghint/index.md

```diff
@@ -0,0 +1,108 @@
+
+The Scheduling Queue is composed of three data structures: ActiveQ, BackoffQ and Unschedulable Pod Pool.
+- ActiveQ: It holds newly created Pods or Pods which are ready to be retried for scheduling.
+- BackoffQ: It holds Pods which are ready to be retried, but are waiting for a backoff period, which depends on the number of times the scheduled attempted to schedule the Pod.
+- Unschedulable Pod Pool: It holds Pods which should not be scheduled for now, because they have a Scheduling Gate or because the scheduler attempted to schedule them and nothing has changed in the cluster that could make the Pod schedulable.
+
+## Scheduling framework and plugins
+
+The Kubernetes scheduler is implemented following the Kubernetes
+[scheduling framework](/docs/concepts/scheduling-eviction/scheduling-framework/).
+
+And, each scheduling requirements are implemented as a plugin.
+(e.g., [Pod affinity](/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity)
+is implemented in the `PodAffinity` plugin.)
+
+The first phase, called the Scheduling Cycle, takes Pods from activeQ **one by one**, runs all plugins' logic,
+and lastly decides in which Node to run the Pod, or concludes that the Pod cannot go to anywhere for now.
+
+If the scheduling is successful, the second phase, called Binding Cycle, binds the Pod with the Node by communicating the decision to the apiserver.
+But, if it turns out that the Pod cannot go to anywhere in Scheduling Cycle, 
+Binding Cycle isn't executed, instead the Pod is moved back to Scheduling Queue.
+Although there are some exceptions, unscheduled Pods enter the Unschedulable Pod Pool.
+
+Pods in Unschedulable Pod Pool are moved to ActiveQ/BackoffQ 
+only when Scheduling Queue identifies changes in the cluster that might be schedulable if we retry the scheduling.
+
+That is a crucial step because Scheduling Cycle is performed for Pods one by one -
+if we didn't have Unschedulable Pod Pool and kept retrying the scheduling of any Pods, 
+multiple Scheduling Cycles would be wasted for Pods with no chance to be scheduled.
+
+Then, how do they decide when to move a Pod back into the ActiveQ? How do they notice that Pods might be schedulable now?
+Here QueueingHints come into play.
+
+## QueueingHint
+
+QueueingHint is callback function per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)
+that may make Pods schedulable.
+
+Let's say the Pod `pod-a` has a required Pod affinity, and got rejected in scheduling cycle by the `PodAffinity` plugin
+because no Node has any Pod matching the Pod affinity specification for `pod-a`.
+
+![pod-a got rejected by PodAffinity](./queueinghint1.svg)
+
+When an unscheduled Pod is put into Unschedulable Pod Pool, Scheduling Queue records which plugins caused the scheduling failure of the Pod.
+In this example, Scheduling Queue notes that PodA was rejected by PodAffinity.
```

Comment:
I'm puzzled. A queue is usually a data structure and not an actor that performs work.

Is “Scheduling Queue” the name of a component within the scheduler than performs action, or a list of Pods awaiting scheduling? I cannot guess.

```suggestion
When an unscheduled Pod is put into the unschedulable pod pool, the scheduling queue
records which plugins caused the scheduling failure of the Pod.
In this example, scheduling queue notes that `pod-a` was rejected by PodAffinity.
```

Reply:
Scheduling Queue is one of the most complicated part in the scheduler. So, it's called queue, but not just a data structure.
So, we call a component in the scheduler "Scheduling Queue" - it has activeQ, backoffQ, and unsched pod pool internally as described in this post, and also performs actions with them (like moving Pods based on cluster events, handling backoff, etc).